WT-10626 Add testing backups to timestamp_abort

pmacko86 opened the pull request:
This PR makes the following changes in relation to WT-10626:

- Refactors database verification into `recover_and_verify()` (sorry for the ugly diff!)
- Adds support for running more work after the database abort and recovery via the new `-i` argument, as some errors may happen only when we are resuming work after a crash
- Adds support for taking full and incremental backups after each checkpoint, enabled via the `-B` argument.
- Adds backup verification, either by reusing the logic of `recover_and_verify()` or by using `testutil_verify_src_backup()` (the latter is currently used by default)

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +363,218 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+                 * If this is the first range, copy the base file from the source backup. Then open
+                 * the source and target files to set up the rest of the copy.
+                 */
+                if (first_range) {
```
Why are we copying with a RANGE type at all? If we have a range we want to just copy it. Using incremental should take a full backup first, which is what you do in `backup_create_full`. So your target files should be there already.

sueloverso replied:
I think a good comment explaining what is going on (and maybe a few smaller comments in the code elsewhere referencing some aspect of the strategy) will be sufficient. Thanks for you explanation.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +363,218 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+    /* Remember that the backup finished successfully. */
+    testutil_check(__wt_snprintf(buf, sizeof(buf), "%s/done", backup_home));
+    testutil_assert((wfd = open(buf, O_WRONLY | O_CREAT, 0666)) >= 0);
+    testutil_assert(close(wfd) == 0);
```
These two functions are very, very similar to the code in `incr_backup`. Could they be refactored into `testutils`?

pmacko86 replied:
My guess is that as described in my earlier comment, this test handles backups differently form incr_backup, it's not immediately clear to me how much refactoring is possible. But I'll look into it.

sueloverso replied:
If not that is fine. Just worth a try initially.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +363,218 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+/*
+ * backup_delete_old_backups --
+ *     Delete old backups, keeping just a few recent ones, so that we don't take too much space for
+ *     no good reason.
```
I don't understand how we have many backups. Can you add a comment somewhere to explain the strategy and algorithm for adding backups here?

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -374,9 +612,30 @@ thread_ckpt_run(void *arg)
+                last_backup = u;
+        }
+        testutil_assert(ret == WT_NOTFOUND);
+        testutil_check(cursor->close(cursor));
```
It may be more useful to have the backup thread separate from checkpoint. It is better coverage to have them active at the same time sometimes.

sueloverso commented on test/utility/misc.c:
```diff
@@ -716,3 +716,27 @@ is_mounted(const char *mount_dir)
+        testutil_die(ERANGE, "The command is too long.");
+
+    testutil_check(system(buf));
+}
```
This feels useful, but I think there are probably a lot of `system()` calls in the test code that should now use this. Maybe make a new ticket to fix that code once this is merged because tests should either use this or we should not have it.

pmacko86 replied:
I agree with this—once this is merged, we can create another ticket to handle this.

sueloverso reviewed (commented):
This is a good way along. You mentioned it wasn't working but no details about was going wrong. I suggest perhaps getting it working one change at a time:
* Add the backup creation thread/code. (Refactor to testutils perhaps)
* Add the backup verify code in the parent.
* Refactor the main code (or maybe split out to another ticket).
* I would suggest leaving the iteration changes to another ticket as I think it unnecessarily complicates this change.

pmacko86 commented:
> This is a good way along. You mentioned it wasn't working but no details about was going wrong.

Here is the failure I've been getting:
```
Verify backup 1 (quick)
transaction: rollback to stable pages visited: 0
transaction: rollback to stable pages visited: 1
cmp: EOF on backup.1/collection.wt which is empty
FAIL: status 256 ID ID1 from cmd: cmp -n 16384 WT_HOME/collection.wt backup.1/collection.wt 375095296 375095296
test_timestamp_abort: FAILED: void testutil_verify_src_backup(WT_CONNECTION *, const char *, const char *, char *)/409: status == 0
test_timestamp_abort: process aborting
WiredTiger Error: void __wt_abort(WT_SESSION_IMPL *), 28: aborting WiredTiger library
Aborted
```

I sent this to you privately earlier, but next time I'll mention it in the ticket or the PR, so that it's all in one place :)

I even got this one, which hints at an error in the middle of a file, which worries me more:
```
transaction: rollback to stable pages visited: 0
WT_HOME/local.wt backup.1/local.wt differ: byte 725001, line 8
FAIL: status 256 ID ID1 from cmd: cmp -n 2031616 WT_HOME/local.wt backup.1/local.wt 25526272 25526272
test_timestamp_abort: FAILED: void testutil_verify_src_backup(WT_CONNECTION *, const char *, const char *, char *)/409: status == 0
test_timestamp_abort: process aborting
WiredTiger Error: void __wt_abort(WT_SESSION_IMPL *), 28: aborting WiredTiger library
Aborted
```

> I suggest perhaps getting it working one change at a time:
> 
> * Add the backup creation thread/code. (Refactor to testutils perhaps)
> * Add the backup verify code in the parent.

> * Refactor the main code (or maybe split out to another ticket).

I had to refactor the recovery and verification code into `recover_and_verify()` so that it is reusable for full backup verification (as an alternative to `testutil_verify_src_backup`), and also so that `main()` is shorter.

To clarify, would it make sense for me to pull this out as a separate ticket that we would merge before these other changes to the test? This would be useful especially if more than one person works on the test (I think Don was planning to make some changes as well?).


> * I would suggest leaving the iteration changes to another ticket as I think it unnecessarily complicates this change.

Yes, this can be a separate ticket.

sueloverso commented:
> I even got this one, which hints at an error in the middle of a file, which worries me more:
> FAIL: status 256 ID ID1 from cmd: cmp -n 2031616 WT_HOME/local.wt backup.1/local.wt 25526272  25526272

This failure looks like it could be what we're looking for in reproducing the error.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -731,18 +1008,19 @@ run_workload(void)
     for (i = 0; i < nth; ++i) {
-        init_thread_data(&td[i], conn, WT_BILLION * (uint64_t)i, i);
+        init_thread_data(
+          &td[i], conn, WT_BILLION * (uint64_t)iteration + WT_MILLION * (uint64_t)i, i, iteration);
```
Thanks. I've used the keyspace contents often (via hexdump) to look at failures so having an easily understood key usage is very important.

pmacko86 commented:
I hope this now addresses the concerns 👍 

The test still, however, fails with:
```
...
transaction: rollback to stable pages visited: 8397
transaction: rollback to stable pages visited: 8465
transaction: rollback to stable pages visited: 8485
Connection open and recovery complete. Verify content
Got stable_val 3439891
records/records-0: Ignore partial record 267051 last valid key 267049
records/records-2: Ignore partial record 2000507461 last valid key 2000507459
transaction: rollback to stable pages visited: 8485
1258467 records verified
Verify backup 3 (quick)
transaction: rollback to stable pages visited: 0
transaction: rollback to stable pages visited: 0
Verify backup 4 (quick)
transaction: rollback to stable pages visited: 0
transaction: rollback to stable pages visited: 0
Verify backup 5 (quick)
transaction: rollback to stable pages visited: 0
transaction: rollback to stable pages visited: 0
transaction: rollback to stable pages visited: 0
cmp: EOF on backup.5/collection.wt which is empty
FAIL: status 256 ID ID5 from cmd: cmp -n 16384 WT_HOME/collection.wt backup.5/collection.wt 742522880 742522880
test_timestamp_abort: FAILED: void testutil_verify_src_backup(WT_CONNECTION *, const char *, const char *, char *)/409: status == 0
test_timestamp_abort: process aborting
WiredTiger Error: void __wt_abort(WT_SESSION_IMPL *), 28: aborting WiredTiger library
Aborted
```

If you enable `backup_verify_throughout` (which is not currently exposed through the command-line arguments), the test fails pretty quickly even before abort, e.g.:
```
Create checkpoint thread
Create timestamp thread
Create backup thread
Create 3 writer threads
Thread 1 starts at 1000000000
Thread 0 starts at 0
Thread 2 starts at 2000000000
Create backup: Type: full, This ID: 1, Consolidate: 1
Checkpoint 1 complete: Flush: NO, at stable 923284.
Create backup: Files: 20
Create backup: Type: incremental, Source ID: 1, This ID: 2
Checkpoint 2 complete: Flush: NO, at stable 1381393.
Create backup: Files: 47, Ranges: 24610, Unmodified: 0
WT_HOME/collection.wt backup.2/collection.wt differ: byte 9, line 1
FAIL: status 256 ID ID2 from cmd: cmp -n 16384 WT_HOME/collection.wt backup.2/collection.wt 299335680 299335680
test_timestamp_abort: FAILED: void testutil_verify_src_backup(WT_CONNECTION *, const char *, const char *, char *)/409: status == 0
test_timestamp_abort: process aborting
WiredTiger Error: void __wt_abort(WT_SESSION_IMPL *), 28: aborting WiredTiger library
test_timestamp_abort: FAILED: Child process 15571 abnormally exited: Invalid argument
test_timestamp_abort: process aborting
WiredTiger Error: void __wt_abort(WT_SESSION_IMPL *), 28: aborting WiredTiger library
Aborted
```

Please let me know if you think this is due to my bug (e.g., using `testutil_verify_src_backup` incorrectly), or if this may be the bug that we are trying to reproduce.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +371,241 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+    testutil_assert((wfd = open(buf, O_WRONLY | O_CREAT, 0666)) >= 0);
+    testutil_assert(close(wfd) == 0);
+
+    printf("Create backup: Files: %" PRId32 ", Ranges: %" PRId32 ", Unmodified: %" PRId32 "\n",
```
I suggest adding "Create incremental backup"

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +371,241 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+    testutil_assert((wfd = open(buf, O_WRONLY | O_CREAT, 0666)) >= 0);
+    testutil_assert(close(wfd) == 0);
+
+    printf("Create backup: Files: %" PRId32 "\n", nfiles);
```
I suggest adding "Create full backup"

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +371,241 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+        if (wfd > 0)
+            testutil_assert(close(wfd) == 0);
+
+        if (first_range) {
```
I think this needs a bigger comment. Something like:
/* 
 * If we get here and first_range is still true it means there there were no changes to the file.
 * The duplicate backup cursor did not return any information. Therefore "copy" the file from
 * the source backup...
 * /

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +371,241 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+        count = 0;
+        while ((dir = readdir(d)) != NULL) {
+            if (strncmp(dir->d_name, "backup.", 7) == 0) {
+                indexes[count++] = atoi(dir->d_name + 7);
```
Minor nit: I suggest making "backup." a global macro since it is used elsewhere and then here:
len = strlen(BACKUP_BASE); instead of "7" in both lines.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +371,241 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+                testutil_check(__wt_snprintf(buf, sizeof(buf), "%s/done", dir->d_name));
+                if (stat(buf, &sb) != 0 && errno == ENOENT) {
+                    testutil_system("%s %s", RM_COMMAND, dir->d_name);
+                }
```
Style: Don't put {} around single statements.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +371,241 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+        __wt_qsort(indexes, (size_t)count, sizeof(*indexes), __int_comparator);
+        for (i = 0; i < count - retain; i++) {
+            testutil_system("%s backup.%d", RM_COMMAND, indexes[i]);
+        }
```
Same style comment. Remove {}

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -374,9 +640,14 @@ thread_ckpt_run(void *arg)
     for (i = 1;; ++i) {
         sleep_time = __wt_random(&td->extra_rnd) % MAX_CKPT_INVL;
         testutil_tiered_sleep(opts, session, sleep_time, &flush_tier);
+
```
Both of these changes to `thread_ckpt_run` no longer seem necessary since you moved the backup code out.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -412,6 +683,48 @@ thread_ckpt_run(void *arg)
+     * Create backups until we get killed.
+     */
+    for (i = 1;; ++i) {
+        sleep_time = __wt_random(&td->extra_rnd) % MAX_CKPT_INVL;
```
I thought about whether we wanted an independent MAX_BACKUP_INVL. But I think this is okay.

sueloverso commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -734,34 +1049,44 @@ run_workload(void)
     /*
      * NOTREACHED
      */
+
```
Remove added blank lines.

sueloverso reviewed (commented):
I have a lot of small suggested changes. I like the look of this diff now and can follow and understand what you did. I'll take one more look after this review.

pmacko86 commented on test/csuite/timestamp_abort/main.c:
```diff
@@ -340,6 +371,241 @@ set_flush_tier_delay(WT_RAND_STATE *rnd)
+    /* Remember that the backup finished successfully. */
+    testutil_check(__wt_snprintf(buf, sizeof(buf), "%s/done", backup_home));
+    testutil_assert((wfd = open(buf, O_WRONLY | O_CREAT, 0666)) >= 0);
+    testutil_assert(close(wfd) == 0);
```
I converted `open/close` to `fopen/fclose` to all cases except the one where we need `pread/pwrite` (similarly as `incr_backup` also uses those two functions). Let me know if there is a better way to do it.

sueloverso commented:
@pmacko86 it is difficult to tell if the failures are reproducing the issue or not. Once this is merged, I will run it and dive into it. Or I can pull the branch itself and run on my box. What is the command to run/reproduce the failures you see?

I think it will be fine to merge because none of the scripts should be using the backup arg. Perhaps we should open a ticket to add new calls into `smoke.sh` but only once we use it for our repro purposes and understand what is going on.

pmacko86 commented:
> @pmacko86 it is difficult to tell if the failures are reproducing the issue or not. Once this is merged, I will run it and dive into it. Or I can pull the branch itself and run on my box. What is the command to run/reproduce the failures you see?

Sounds good. I suppose there is no harm in merging?

I just reproduce these problems with `test_timestamp_abort -B -T 3 -t 20` (3 threads, 20 seconds).

sueloverso reviewed (approved):
LGTM
